### PR TITLE
New MRS v2 cluster resource

### LINKED ISSUE #1149: Plan for MRS v2
Now the mrs cluster resource is using v1.1 API, not the latest v2 API.
For that reason, some features are not supported. That contains:
MIXED and CUSTOM types of cluster
security group of cluster
public ip of cluster
enterprise_project_id of of cluster
configurable type and size of master node disk
configurable type, size and number of core node disk
agency of cluster
Flink type of job

We should plan to upgrade the mrs resource to v2

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
Compared with the V1 version of the cluster, the V2 version have a more user-friendly parameter structure, and supports enterprise project ID and node expansion and contraction.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1149 #1306 

**Special notes for your reviewer**:

**Release note**:

```release-note
1. New cluster resource support:
    a. support master, core and task nodes management
    b. support resize core groups for core and task group
2. Related document support.
3. Related acc test support (except for the failure of the ANALYSIS cluster due to the security group rule port 9022 error).
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed
basic test
```
make testacc
TEST='./huaweicloud/services/acceptance/mrs' TESTARGS='-run=TestAccMrsMapReduceCluster_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/mrs -v -run=TestAccMrsMapReduceCluster_basic -timeout 360m -parallel 4
=== RUN   TestAccMrsMapReduceCluster_basic
=== PAUSE TestAccMrsMapReduceCluster_basic
=== CONT  TestAccMrsMapReduceCluster_basic
--- PASS: TestAccMrsMapReduceCluster_basic (1352.10s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/mrs1352.177s
```
with keypair
```
make testacc 
TEST='./huaweicloud/services/acceptance/mrs' TESTARGS='-run=TestAccMrsMapReduceCluster_keypair'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/mrs -v -run=TestAccMrsMapReduceCluster_keypair -timeout 360m -parallel 4
=== RUN   TestAccMrsMapReduceCluster_keypair
=== PAUSE TestAccMrsMapReduceCluster_keypair
=== CONT  TestAccMrsMapReduceCluster_keypair
--- PASS: TestAccMrsMapReduceCluster_keypair (727.97s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/mrs 728.042s
```

### CODE REVIEW

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go [hunk 1]
```diff
@@ -0,0 +1,906 @@
...
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/huaweicloud/golangsdk"
+	"github.com/huaweicloud/golangsdk/openstack/common/tags"
+	"github.com/huaweicloud/golangsdk/openstack/mrs/v1/cluster"
+	"github.com/huaweicloud/golangsdk/openstack/mrs/v2/clusters"
```
Comment (ShiChangkuo) [inline huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: as we used both v1 and v2 packages, suggest to give an alias for one of them

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go [hunk 2]
```diff
@@ -0,0 +1,906 @@
...
+					typeAnalysis, typeStream, typeHybrid,
+				}, false),
+			},
+			"enterprise_project_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"ip_address": {
```
Comment (ShiChangkuo) [inline huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: suggest to use `eip_id` instead.

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go [hunk 3]
```diff
@@ -0,0 +1,906 @@
...
+			},
+			"total_node_number": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"master_node_ip": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"external_ip": {
```
Comment (ShiChangkuo) [inline huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: public_ip?

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go [hunk 4]
```diff
@@ -0,0 +1,906 @@
...
+	}
+	if v, ok := d.GetOk("node_key_pair"); ok {
+		createOpts.NodeKeypair = v.(string)
+		createOpts.LoginMode = "KEYPAIR"
+	} else {
+		createOpts.NodeRootPassword = d.Get("node_admin_pass").(string)
+		createOpts.LoginMode = "PASSWORD"
+	}
+
+	mrsV2Client, err := config.MrsV2Client(config.GetRegion(d))
```
Comment (ShiChangkuo) [inline huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: we always get the service clients at the head of one function.

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go [hunk 5]
```diff
@@ -0,0 +1,906 @@
...
+		return fmtp.Errorf("Error creating HuaweiCloud MRS client: %s", err)
+	}
+	if d.HasChange("tags") {
+		tagErr := utils.UpdateResourceTags(client, d, "clusters", d.Id())
+		if tagErr != nil {
+			return fmtp.Errorf("Error updating tags of MRS cluster:%s, err:%s", d.Id(), tagErr)
+		}
+	}
+	// lintignore:R005
+	if d.HasChange("analysis_core_nodes") || d.HasChange("streaming_core_nodes") ||
```
Comment (ShiChangkuo) [inline huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: can we use d.HasChanges ?

File: huaweicloud/services/acceptance/mrs/resource_huaweicloud_mapreduce_cluster_test.go [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+%s
+
+resource "huaweicloud_compute_keypair" "test" {
+  name = "%s"
+
+  lifecycle {
+    ignore_changes = [
+      public_key,
+    ]
+  }
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/mrs/resource_huaweicloud_mapreduce_cluster_test.go:482]: why we add `lifecycle` block here?

File: docs/resources/mapreduce_cluster.md [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) The region in which to create the MapReduce cluster resource.
+  If omitted, the provider-level region will be used.
+  Changing this will create a new MapReduce cluster resource.
+
+* `availability_zone` - (Required, String, ForceNew) Specifies the availability zone in which to create
+  the cluster. Please following [reference](https://developer.huaweicloud.com/en-us/endpoint/?ECS)
```
Comment (ShiChangkuo) [inline docs/resources/mapreduce_cluster.md]: the link of `reference` is not correct?

File: docs/resources/mapreduce_cluster.md [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+  Changing this will create a new MapReduce cluster resource.
+
+* `name` - (Required, String, ForceNew) Specifies the name of the MapReduce cluster.
+  The name can contain 2 to 64 characters, which may consist of letters, digits,
+  underscores (_) and hyphens (-).
+  Changing this will create a new MapReduce cluster resource.
+
+* `version` - (Required, String, ForceNew) Specifies the MapReduce cluster version.
+  Changing this will create a new MapReduce cluster resource.
```
Comment (ShiChangkuo) [inline docs/resources/mapreduce_cluster.md:204]: suggest to add an example

File: docs/resources/mapreduce_cluster.md [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+  The `nodes` object structure of the `master_nodes` is documented below.
+  Changing this will create a new MapReduce cluster resource.
+
+* `vpc_id` - (Required, String, ForceNew) Specifies the VPC ID.
+  Changing this will create a new MapReduce cluster resource.
+
+* `subnet_id` - (Required, String, ForceNew) Specifies the network ID of a subnet.
+  Changing this will create a new MapReduce cluster resource.
+
+  -> **NOTE:** The duplication of subnet CIDR will cause the MRS cluster creation to fail.
```
Comment (ShiChangkuo) [inline docs/resources/mapreduce_cluster.md]: can we create a subent with duplication CIDR?

File: docs/resources/mapreduce_cluster.md [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The cluster ID in UUID format.
+* `total_node_number` - The total number of nodes deployed in a cluster.
+* `master_node_ip` - The IP address of a master node.
+* `external_ip` - The external IP address.
+* `private_ip` - The preferred private IP address.
+* `internal_ip` - The internal IP address.
+* `external_alternate_ip` - IP address of a Master node.
```
Comment (ShiChangkuo) [inline docs/resources/mapreduce_cluster.md]: what's the differents between xx_ip ?

File: docs/resources/mapreduce_cluster.md [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+
+* `data_volume_type` - (Optional, String, ForceNew) Specifies the data disk flavor of the nodes.
+  Required if `data_volume_count` is not empty.
+  Changing this will create a new MapReduce cluster resource.
+
+* `data_volume_size` - (Optional, Int, ForceNew) Specifies the data disk size of the nodes.
+  Required if `data_volume_count` is not empty.
+  Changing this will create a new MapReduce cluster resource.
+
+* `data_volume_count` - (Optional, Int, ForceNew) Specifies the data disk number of the nodes.
```
Comment (ShiChangkuo) [inline docs/resources/mapreduce_cluster.md:303]: please add more description of the parameter.

**Review by ShiChangkuo — CHANGES_REQUESTED** [review]

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: The v2 sdk use `clusterV2` as an alias now.

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: Has been replaced.
The eip_id cannot be find by VPC and MRS API, so the value is not set to state.

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: Not public IP, it's a external IP for MRS managent page and the `external_ip` has been removed.

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: All client are moved to the place of the function header.

File: huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/services/mrs/resource_huaweicloud_mapreduce_cluster.go]: Has been updated

File: docs/resources/mapreduce_cluster.md (hunk 7 above)
Comment (Lance52259) [reply docs/resources/mapreduce_cluster.md]: the link has been updated with MRS reference.

File: docs/resources/mapreduce_cluster.md (hunk 9 above)
Comment (Lance52259) [reply docs/resources/mapreduce_cluster.md]: the description has been removed.

File: docs/resources/mapreduce_cluster.md (hunk 10 above)
Comment (Lance52259) [reply docs/resources/mapreduce_cluster.md]: The `private_ip` is the virtual ip of the HA master node of the master node group.
The `external_ip`, `external_alternate_ip` and `internal_ip` are not used, has been removed.

File: docs/resources/mapreduce_cluster.md (hunk 11 above)
Comment (Lance52259) [reply docs/resources/mapreduce_cluster.md:303]: The limit of the `data_volume_count` for each node groups has been supported.

File: huaweicloud/services/acceptance/mrs/resource_huaweicloud_mapreduce_cluster_test.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/services/acceptance/mrs/resource_huaweicloud_mapreduce_cluster_test.go:482]: after applying terraform configuration, the plan will be not empty (public_key is a ForceNew parameter).

File: docs/resources/mapreduce_cluster.md (hunk 8 above)
Comment (Lance52259) [reply docs/resources/mapreduce_cluster.md:204]: add support desctiption for version `MRS 1.9.2`, version `MRS 3.0.5` and version `MRS 3.1.0`.

**Review by ShiChangkuo — APPROVED** [review]

**Pull request merged**